### ISSUE TITLE
Response body should be able to be read after fetching

### ISSUE THREAD

**Issue by Raiondesu** [drwpow/openapi-typescript#1159]
**Description**

Currently, if one attempts to read a response body in any way (`.text()`/`.blob()`/etc.) after fetching, they will be presented with the following error:
```
TypeError: Failed to execute ['blob'/'text'/etc.] on 'Response': body stream already read
```

**Reproduction**

1. Set up any openapi-fetch client:
```ts
import createClient from 'openapi-fetch';
import type { paths } from './schema'; // generated via openapi-typecript

const client = createClient<paths>({ /* any config */ });
```
2. Call `json()` or `blob()` in a `then` block after the request:
```ts
client
  .get(''/* any request */, { /* any requestinit */ })
  .then(r => r.response.blob()) // <-- throws TypeError
```
3. Run and observe the console:
```
TypeError: Failed to execute 'blob' on 'Response': body stream already read
```

**Expected result**

I argue the user should expect the `response` property to contain a fully functional original response, with the ability to parse the body without any errors in any way should the need arise.

**Checklist**

- [x] I’m willing to open a PR, as I have already implemented a fix using [response.clone()](https://developer.mozilla.org/en-US/docs/Web/API/Response/clone)

**Comment by drwpow** [comment]
Question: why would you want to get `.text()` for a typed fetch library for a JSON API? If your schema doesn’t list a JSON response, this library will automatically call `.text()` for you.

Trying to understand the real-world usecase here to provide a good workaround. I’d like to not have to require everyone calling `.json()` manually themselves for 99.9% of usecases; IMO that’s a bad API that leads to noisy boilerplate (fine when using the fetch API, but not for a wrapper meant for JSON calls)

**Comment by drwpow** [comment]
> My usecase (and I assume it's possible I'm not the only one) is to read the `blob` from the response body.

Could you give more specific detail why you’re trying to read a `blob` for an endpoint marked as JSON in your schema? I’m looking for any clue your schema would give me that you would want to do this.

As you’re probably already aware, that error comes directly from the native fetch API. because we call `json()` internally, you can’t then call `blob()` afterward. Same as the normal fetch API. So internally, if we know that an endpoint has no `application/json` content type, this library calls `.text()` instead. We could add `blob()` support as well based off content-type hints, but that’s currently a bit of a can of worms.

Anyways, all that said, I’d like to assume that **your OpenAPI schema contains all the information needed to infer which response you want** because that’s the whole point of this library. In cases where that’s not possible, though, perhaps this could add an option to override the automatic `.json()` call. But in exchange for getting a `string` or `ArrayBuffer` type back instead (which I’d also like to know which benefit this library provides for you in those cases, so this can be most helpful)

**Comment by Raiondesu** [comment]
I believe you have misunderstood my intentions with the implementation a little. I'll try to be a bit more specific this time.
I propose a simple ~15-line fix, without breaking any currently established logic **except** getting the native `fetch` error.

Please, take a look: https://github.com/Raiondesu/openapi-typescript/commit/92f0412c03bfa11ef51db6c517ae3d44ab62e37e

---

> I’m also not interested in not calling the fetch API’s built-in response.json() method [...]

I understand this and I do agree with you, so this is not what I propose. My proposal involves simply cloning the request before calling `.json()`. This makes it so the actual response the user gets is indistinguishable from the native `fetch` response (`body` is unused), while not sacrificing anything - the user still gets the (possibly) parsed json in the `data` property exactly as they do now.

> [...] I’d love additional detail as to why you’d want a blob() for a JSON response

Besides that the details of my usecase are under an NDA, I believe that they will not be of any value to this discussion anyway, as I am certain that my proposal improves user-experience of the library in general when processing a response, rather than addressing a highly-specific usecase.

---

P.S.
> So internally, if we know that an endpoint has no `application/json` content type, this library calls `.text()` instead.

Sorry for the trouble, could you please reference where exactly the library does this?\
I'm feeling a bit lost about this: I have searched for any mention of `.text()` in the code and couldn't find any, so I feel like I must be missing something obvious. I could only find the `.json()` being called [here](https://github.com/drwpow/openapi-typescript/blob/3a4f410beb9af73c48690c3826c0b9ef8002dbab/packages/openapi-fetch/src/index.ts#L95).

**Comment by drwpow** [comment]
Thanks for providing more detail, and a great suggestion on implementation. I do think solving this is valuable, as it relates to #1123 and #1157.

I’ve added a new `parseAs` option in #1169 that allows you to override the default `.json()` handling if desired.

> could you please reference where exactly the library does this?

I apologize; it doesn’t! I was getting confused with some in-flight work / other conversations where we had talked about doing it. But it never ended up landing. But all that will change in the next release.

**Comment by drwpow** [comment]
This is fixed in `0.3.0`. Thanks again!
